compiler/src/dmd/location.d

```diff
@@ -60,19 +64,45 @@ nothrow:
         this.messageStyle = messageStyle;
     }
 
-    extern (D) this(const(char)* filename, uint linnum, uint charnum) pure
+    extern (D) this(const(char)* filename, uint linnum, uint charnum)
     {
         this.linnum = linnum;
-        this.charnum = charnum;
+        this.charnum = cast(ushort)charnum;
         this.filename = filename;
     }
 
+    /***
+     * Returns: filename for this location, null if none
+     */
+    extern (C++) const(char)* filename() const @nogc
+    {
+        return fileIndex ? filenames[fileIndex - 1] : null;
+    }
+
+    /***
+     * Set file name for this location
+     * Params:
+     *   name = file name for location, null for no file name
+     */
+    extern (C++) void filename(const(char)* name)
+    {
+        if (name)
+        {
+            //printf("setting %s\n", name);
+            filenames.push(name);
+            fileIndex = cast(ushort)filenames.length;
+            assert(fileIndex);  // no overflow
```

Comment:
Won't this result in many duplicates?

Reply:
Ironically, no, for the same reason the lexer doesn't lex the same file multiple times.